clustering/ejb/infinispan/src/main/java/org/wildfly/clustering/ejb/infinispan/ExpirationTracker.java

```diff
@@ -0,0 +1,254 @@
+
+/**
+ * Tracks expiration of multiple id objects.
+ *
+ * @author Flavia Rainone
+ */
+class ExpirationTracker<I> {
+
+    // the expiration timeout in milisseconds
+    private final long timeout;
+    // all expiration infos being tracked mapped by id
+    private final Map<I, ExpirationInfo> trackedIds;
+    // the first expiration to occur
+    private ExpirationInfo firstExpiration;
+    // the last expiration to occur
+    private ExpirationInfo lastExpiration;
+    // expiration info: a single node in a double linked list
+    private class ExpirationInfo {
+        ExpirationInfo nextExpiration;
+        ExpirationInfo previousExpiration;
+        volatile long expiration;
+        final I id;
+
+        ExpirationInfo(I id) {
+            this.id = id;
+        }
+    }
+
+    /**
+     * Creates an ExpirationTracker with configured {@code timeout} for expiration
+     * @param timeout the time it takes for an id to expire.
+     */
+    ExpirationTracker(Duration timeout) {
+        this.timeout = timeout.toMillis();
+        trackedIds = new HashMap<>();
+    }
+
+    /**
+     * Tracks expiration of bean identified by {@code id} starting the count down
+     * now.
+     *
+     * @param id the id of the bean that will expire
+     */
+    void trackExpiration(I id) {
+        // TODO add log messages for tracing?
+        final ExpirationInfo expiration;
+        synchronized (this) {
+            // if we are already tracking this id, move the id to the lastExpiration
+            // position, as it will be the last id to expire;
+            if (trackedIds.containsKey(id)) {
+                expiration = trackedIds.get(id);
+                // remove the node from the expiration linked list
+                if (expiration.nextExpiration != null) {
+                    expiration.nextExpiration.previousExpiration = expiration.previousExpiration;
+                }
+                if (expiration.previousExpiration != null) {
+                    expiration.previousExpiration.nextExpiration = expiration.nextExpiration;
+                } else if (firstExpiration == expiration &&
+                        firstExpiration != lastExpiration) {
+                    firstExpiration = firstExpiration.nextExpiration;
+                }
+                expiration.previousExpiration = null;
+                expiration.nextExpiration = null;
+            } else {
+                // if we are not tracking the id at all, add it to the expiration list
+                expiration = new ExpirationInfo(id);
+                trackedIds.put(id, expiration);
+            }
+            // if the list is empty, now add a single expiration node
+            if (lastExpiration == null) {
+                firstExpiration = lastExpiration = expiration;
+            } else if (lastExpiration != expiration) {
+                // if the last expiration to occur is not current node, add it to the
+                // last of the expiration list
```

Comment:
This logic makes the assumption that the newly tracked bean will be the last to expire.  While this is true when scheduling a bean for expiration at the end of an invocation, this is generally not the case when a member takes on the responsibility for expiring a given bean, i.e. when triggered from InfinispanBeanManager.schedule(Locality oldLocality, Locality newLocality).

Reply:
Oh, I see. I'll try to think on a solution and I'll assert the perf gains with the benchmark to make sure this PR is really needed.

Reply:
One solution is to replace the expiration tracker with a collection, e.g. io.undertow.util.ConcurrentDirectDeque.  We can then construct the scheduler with a queue implementation appropriate for the use case.  In the clustered case, where the assumption above is violated, we could use a priority queue, where elements are inserted according to their next expiration, rather then just appending to the end.  Insertions/removals would be slower log(N) instead of constant time, but that way we can still optimize for the local case.